### DISCUSSION TITLE
Jacobian support

### DISCUSSION THREAD

**Discussion by zxhuang97** [NVlabs/curobo#330] [Software Q&A] [answered]
Hi,

Thank you for creating this wonderful library.

I noticed that calculate_jacobian is currently disabled for the forward kinematics. Is there a timeline for when this feature will be available? I imagine the Jacobian is computed and used in IK, so it might just need to be exposed to the Python side. Am I understanding this correctly?

Thanks!

------------------------------- Update ----------------------------
Based on #150, I guess it's not as easy as I thought.

**Comment by balakumar-s** [comment]
Yes, it's not easy yet. We did think of adding it but haven't figure out how best to implement it. What use case do you have in mind for using Jacobian?

**Reply by zxhuang97** [reply]
Jacobian is crucial in contact-rich manipulation. For example, we need Jacobian to compute the corresponding torque required to apply a specific force to an object. Additionally, constraints in trajectory optimization (traj opt) are often defined by the Jacobian, such as an equality constraint that keeps a fingertip in contact with an object's surface. We also need jacobian to compute the manipulability cost. In some scenarios, we may also need the gradients of the Jacobian.

An easy way to obtain the Jacobian is to back-propagate through the FK, but this approach can be slow, and computing the gradient of the Jacobian (a second-order gradient) can be complex and cumbersome.

Our lab maintains a [pytorch-based implementation](https://github.com/UM-ARM-Lab/pytorch_kinematics/blob/9074d3bee74848016a8cee484bac236bee6e194f/src/pytorch_kinematics/jacobian.py#L6) for computing  jacobian analytically. While I'm not sure how optimized it is, it might be something to start with. If you are busy with other things, maybe I can take a look at how to translate the code. Have to say that I'm not super familiar with cuda progamming so any advice is appreciated!

**Comment by ommmid** [comment]
@balakumar-s  So none of the internal algorithms in Curobo use Jacobian? I thought it was calculated internally in curobo at some level, and it is just a matter of exposing it. There is a `null_space_weight` parameter in yml file when creating iksolver for example. That lead me to believe that curobo handles nullspace projection that must use Jacobian, can you clarify on that please?

**Comment by balakumar-s** [comment, marked as answer]
We just released a version that outputs kinematic jacobians. See here: https://nvlabs.github.io/curobo/latest/getting-started/forward_kinematics.html
